forc-plugins/forc-client/src/ops/tx_util.rs

```diff
@@ -0,0 +1,181 @@
+            utxo_id: fuel_tx::UtxoId::new(fuel_tx::Bytes32::zeroed(), 0),
+            balance_root: fuel_tx::Bytes32::zeroed(),
+            state_root: fuel_tx::Bytes32::zeroed(),
+            tx_pointer: fuel_tx::TxPointer::new(0, 0),
+        })
+        .add_output(fuel_tx::Output::Contract {
+            input_index,
+            balance_root: fuel_tx::Bytes32::zeroed(),
+            state_root: fuel_tx::Bytes32::zeroed(),
+        })
+    }
+    fn add_contracts(&mut self, contract_ids: Vec<ContractId>) -> &mut Self {
+        for contract_id in contract_ids {
+            self.add_contract(contract_id);
+        }
+        self
+    }
+    fn add_inputs(&mut self, inputs: Vec<Input>) -> &mut Self {
+        for input in inputs {
+            self.add_input(input);
+        }
+        self
+    }
+    async fn fund(
+        &mut self,
+        address: Address,
+        provider: Provider,
+        signature_witness_index: u8,
+    ) -> Result<&mut Self> {
+        let wallet = Wallet::from_address(Bech32Address::from(address), Some(provider));
+
+        let amount = 1_000_000;
+        let asset_id = BASE_ASSET_ID;
+        let inputs = wallet
+            .get_asset_inputs_for_amount(asset_id, amount, signature_witness_index)
+            .await?;
+        let output = Output::change(address, 0, asset_id);
+
+        self.add_inputs(inputs).add_output(output);
+
+        Ok(self)
+    }
+    async fn finalize_signed(
+        &mut self,
+        client: FuelClient,
+        unsigned: bool,
+        signing_key: Option<SecretKey>,
+    ) -> Result<Transaction> {
+        let mut signature_witness_index = 0u8;
+        if !unsigned {
+            // Get the address
+            let address = if let Some(signing_key) = signing_key {
+                Address::from(*signing_key.public_key().hash())
+            } else {
+                prompt_address()?
+            };
+
+            // Insert dummy witness for signature
+            signature_witness_index = self.witnesses().len().try_into()?;
+            self.add_witness(Witness::default());
+
+            // Add input coin and output change
+            self.fund(address, Provider::new(client), signature_witness_index)
+                .await?;
+        }
+
+        let mut tx = self.finalize();
+
+        if !unsigned {
+            let message = Message::new(tx.to_bytes());
+            let signature = if let Some(signing_key) = signing_key {
+                Signature::sign(&signing_key, &message)
+            } else {
+                prompt_signature(message)?
+            };
+
+            let witness = Witness::from(signature.as_ref());
+            tx.replace_witness(signature_witness_index, witness);
+        }
+
+        Ok(tx)
+    }
+}
+
+pub trait TransactionExt {
+    fn replace_witness(&mut self, witness_index: u8, witness: Witness) -> &mut Self;
+}
+
+impl TransactionExt for Transaction {
+    fn replace_witness(&mut self, index: u8, witness: Witness) -> &mut Self {
```

Comment:
Perhaps `insert_witness` could be a better name for this, as it looks like this inserts a witness at the given index?

Reply:
Actually changed the behavior because replacement is needed here.